### DISCUSSION TITLE
How to get schema name for entity in a custom Index?

### DISCUSSION THREAD

**Discussion by linkthai** [mikro-orm/mikro-orm#5109] [Q&A] [answered]
I have a SearchIndex that implements a trgm index for search query.
``` ts 
import { Index } from '@mikro-orm/core';
import { NamingStrategy } from './base.entity';

export const LIKE_INDEX_FN = `lcu`; // lower concat unaccent
export interface SearchIndexOptions {
  type?: 'gin_trgm';
}
export function SearchIndex(options?: SearchIndexOptions) {
  return (target: Partial<any>, propertyName?: string) => {
    const columnName = NamingStrategy.propertyToColumnName(propertyName!);
    const entityName = NamingStrategy.classToTableName(target.constructor.name);

    const { type = 'gin_trgm' } = options || {};

    const name = `${entityName}_${columnName}_${type}`;

    const indexColumn = `${LIKE_INDEX_FN}(${columnName})`;

    const expression = `CREATE INDEX ${name} ON "${entityName}" ${
      type == 'gin_trgm' ? `USING gin` : ``
    }(${indexColumn} ${type}_ops)`;

    Index({ name, expression })(target, propertyName);
  };
}
```
The problem is `ON "${entityName}"` only works for table in public schema, in a specific table schema, I have to write `ON "${schema}".${entityName}`. The problem is I don't have any information about which schema the current target is in, I can't getmetadata as this happens during mikro orm init.

**Comment by B4nan** [comment, marked as answer]
You might have to make this explicit and add a parameter for it, I don't see any other way to get the schema name inside the decorator implementation - mainly because property decorators are executed before class decorators (so before the `@Entity` decorator). Also, the schema might come even later, from the ORM config, and that is evaluated after all the decorators.

Another way in v6 would be trying to leverage the `onMetadata` discovery hook, where you should have all the information needed:

https://mikro-orm.io/docs/configuration#onmetadata-hook

**Reply by linkthai** [reply]
@B4nan After trying the `autodiscovered` hook, it always happens after the Custom Index, so it's out of the question. I will resort to adding an explicit parameter. Thanks for the help.

**Reply by B4nan** [reply]
That's exactly my point, you want something that happens _after_, so you can adjust the expression (regenerate it again) - doing this inside a decorator is a wrong idea, nothing in the ORM works this way, decorators only gather metadata which are later processed in the discovery phase.
